File: ruby/fenwick_Tree.rb
class FenwickTree
  def initialize(size)
    @n = size
    @bit = Array.new(size + 1, 0)
  end

  # Add 'val' to index 'i' (0-based)
  def update(i, val)
    i += 1
    while i <= @n
      @bit[i] += val
      i += i & -i
    end
  end

  # Get prefix sum from 0 to i (0-based)
  def query(i)
    i += 1
    sum = 0
    while i > 0
      sum += @bit[i]
      i -= i & -i
    end
    sum
  end

  # Get sum in range [l, r] (0-based)
  def range_sum(l, r)
    query(r) - query(l - 1)
  end
end

# Example usage:
ft = FenwickTree.new(10)
ft.update(3, 5)     # Add 5 at index 3
ft.update(5, 2)     # Add 2 at index 5
puts ft.query(5)    # Sum from index 0 to 5 → Output: 7
puts ft.range_sum(3, 5)  # Sum from index 3 to 5 → Output: 7

File: c++/fenwick_tree.cpp
// Fenwick Tree (Binary Indexed Tree) - point update, prefix sum
#include <bits/stdc++.h>
using namespace std;

struct Fenwick {
    int n; vector<long long> f;
    Fenwick(int n=0){init(n);}    
    void init(int n_){n=n_; f.assign(n+1,0);}    
    void add(int i,long long delta){ for(; i<=n; i+=i&-i) f[i]+=delta; }
    long long sum(int i){ long long s=0; for(; i>0; i-=i&-i) s+=f[i]; return s; }
    long long range_sum(int l,int r){ if(r<l) return 0; return sum(r)-sum(l-1); }
};

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    // Example: maintain an array of size 10
    Fenwick bit(10);
    bit.add(1,5);
    bit.add(3,2);
    bit.add(7,10);
    cout << "prefix sum up to 7 = " << bit.sum(7) << '\n';
    cout << "range sum 2..7 = " << bit.range_sum(2,7) << '\n';
    return 0;
}
